File: clients/python/tests/stress_concurrent.py
"""LightKV Python Client Stress Test (Concurrent)."""

import time
import sys
import os
import threading
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'lightkv'))

from client import LightKVClient


def worker(thread_id, ops_per_thread, results, lock):
    client = LightKVClient(host='127.0.0.1', port=16379)
    if not client.connect():
        with lock:
            results['failed'] += ops_per_thread * 3
        return

    success = 0
    failed = 0
    latencies = []

    for i in range(ops_per_thread):
        key = f'py_stress_{thread_id}_{i}'
        value = f'value_{i}'

        start = time.time()
        ok = client.set(key, value)
        val = client.get(key)
        client.delete(key)
        elapsed = (time.time() - start) * 1000

        if ok:
            success += 3
        else:
            failed += 3
        latencies.append(elapsed)

    client.quit()

    with lock:
        results['success'] += success
        results['failed'] += failed
        results['latencies'].extend(latencies)


def main():
    num_threads = int(sys.argv[1]) if len(sys.argv) > 1 else 10
    ops_per_thread = int(sys.argv[2]) if len(sys.argv) > 2 else 1000
    total_ops = num_threads * ops_per_thread * 3

    print(f'\n[Python Stress Test]')
    print(f'  Threads: {num_threads}')
    print(f'  Ops/thread: {ops_per_thread} (SET+GET+DELETE)')
    print(f'  Total ops: {total_ops}')

    results = {'success': 0, 'failed': 0, 'latencies': []}
    lock = threading.Lock()
    threads = []

    start = time.time()
    for i in range(num_threads):
        t = threading.Thread(target=worker, args=(i, ops_per_thread, results, lock))
        t.start()
        threads.append(t)

    for t in threads:
        t.join()

    duration_ms = (time.time() - start) * 1000
    latencies = sorted(results['latencies'])
    p50 = latencies[len(latencies) * 50 // 100] if latencies else 0
    p99 = latencies[len(latencies) * 99 // 100] if latencies else 0
    avg = sum(latencies) / len(latencies) if latencies else 0
    ops_sec = results['success'] / (duration_ms / 1000.0)

    print(f'\n  Results:')
    print(f'    Duration:     {duration_ms:.1f} ms')
    print(f'    Success:      {results["success"]} ops')
    print(f'    Failed:       {results["failed"]} ops')
    print(f'    Throughput:   {ops_sec:.0f} ops/sec')
    print(f'    Avg Latency:  {avg:.2f} ms')
    print(f'    P50 Latency:  {p50:.2f} ms')
    print(f'    P99 Latency:  {p99:.2f} ms')
    print(f'\n[Python Stress Test Complete]')


if __name__ == '__main__':
    main()

File: clients/nodejs/test/stress_concurrent.js
const LightKVClient = require('../src/client');

async function worker(threadId, opsPerThread, results) {
    const client = new LightKVClient({ host: '127.0.0.1', port: 16379 });
    await client.connect();

    let success = 0;
    let failed = 0;
    const latencies = [];

    for (let i = 0; i < opsPerThread; i++) {
        const key = `node_stress_${threadId}_${i}`;
        const value = `value_${i}`;

        const start = Date.now();
        const ok = await client.set(key, value);
        await client.get(key);
        await client.delete(key);
        const elapsed = Date.now() - start;

        if (ok) success += 3;
        else failed += 3;
        latencies.push(elapsed);
    }

    await client.quit();
    return { success, failed, latencies };
}

async function main() {
    const numThreads = parseInt(process.argv[2]) || 10;
    const opsPerThread = parseInt(process.argv[3]) || 1000;
    const totalOps = numThreads * opsPerThread * 3;

    console.log('\n[Node.js Stress Test]');
    console.log(`  Threads: ${numThreads}`);
    console.log(`  Ops/thread: ${opsPerThread} (SET+GET+DELETE)`);
    console.log(`  Total ops: ${totalOps}`);

    const start = Date.now();
    const promises = [];
    for (let i = 0; i < numThreads; i++) {
        promises.push(worker(i, opsPerThread));
    }

    const workerResults = await Promise.all(promises);
    const durationMs = Date.now() - start;

    let totalSuccess = 0;
    let totalFailed = 0;
    const allLatencies = [];
    for (const r of workerResults) {
        totalSuccess += r.success;
        totalFailed += r.failed;
        allLatencies.push(...r.latencies);
    }

    allLatencies.sort((a, b) => a - b);
    const p50 = allLatencies[Math.floor(allLatencies.length * 0.5)] || 0;
    const p99 = allLatencies[Math.floor(allLatencies.length * 0.99)] || 0;
    const avg = allLatencies.reduce((a, b) => a + b, 0) / allLatencies.length || 0;
    const opsSec = totalSuccess / (durationMs / 1000);

    console.log('\n  Results:');
    console.log(`    Duration:     ${durationMs.toFixed(1)} ms`);
    console.log(`    Success:      ${totalSuccess} ops`);
    console.log(`    Failed:       ${totalFailed} ops`);
    console.log(`    Throughput:   ${opsSec.toFixed(0)} ops/sec`);
    console.log(`    Avg Latency:  ${avg.toFixed(2)} ms`);
    console.log(`    P50 Latency:  ${p50.toFixed(2)} ms`);
    console.log(`    P99 Latency:  ${p99.toFixed(2)} ms`);
    console.log('\n[Node.js Stress Test Complete]');
}

main().catch(err => {
    console.error('Error:', err);
    process.exit(1);
});
